L1Trigger/CSCTriggerPrimitives/src/CSCGEMMotherboard.cc

```diff
@@ -0,0 +1,349 @@
+                                     unsigned chamber,
+                                     const edm::ParameterSet& conf) :
+  CSCUpgradeMotherboard(endcap, station, sector, subsector, chamber, conf)
+{
+  gemId = GEMDetId(theRegion, 1, theStation, 1, theChamber, 0).rawId();
+  
+  const edm::ParameterSet coPadParams(conf.getParameter<edm::ParameterSet>("copadParam"));
+  coPadProcessor.reset( new GEMCoPadProcessor(endcap, station, 1, chamber, coPadParams) );
+
+  // use "old" or "new" dataformat for integrated LCTs?
+  useOldLCTDataFormat_ = tmbParams_.getParameter<bool>("useOldLCTDataFormat");
+
+  // LCT ghostbusting
+  doLCTGhostBustingWithGEMs_ = tmbParams_.getParameter<bool>("doLCTGhostBustingWithGEMs");
+
+  // promote ALCT-GEM pattern
+  promoteALCTGEMpattern_ = tmbParams_.getParameter<bool>("promoteALCTGEMpattern");
+  promoteALCTGEMquality_ = tmbParams_.getParameter<bool>("promoteALCTGEMquality");
+
+  //  deltas used to match to GEM pads
+  maxDeltaBXPad_ = tmbParams_.getParameter<int>("maxDeltaBXPad");
+  maxDeltaBXCoPad_ = tmbParams_.getParameter<int>("maxDeltaBXCoPad");
+
+  maxDeltaPadL1_ = (par ? tmbParams_.getParameter<int>("maxDeltaPadL1Even") :
+		    tmbParams_.getParameter<int>("maxDeltaPadL1Odd") );
+  maxDeltaPadL2_ = (par ? tmbParams_.getParameter<int>("maxDeltaPadL2Even") :
+		    tmbParams_.getParameter<int>("maxDeltaPadL2Odd") );
+}
+
+CSCGEMMotherboard::CSCGEMMotherboard() : CSCUpgradeMotherboard()
+{
+}
+
+CSCGEMMotherboard::~CSCGEMMotherboard()
+{
+}
+
+void CSCGEMMotherboard::clear()
+{
+  pads_.clear();
+  coPads_.clear();    
+}  
+
+
+void CSCGEMMotherboard::run(const CSCWireDigiCollection* wiredc,
+			    const CSCComparatorDigiCollection* compdc,
+			    const GEMPadDigiClusterCollection* gemClusters)
+{
+  std::unique_ptr<GEMPadDigiCollection> gemPads(new GEMPadDigiCollection());
+  coPadProcessor->declusterize(gemClusters, *gemPads);
+  run(wiredc, compdc, gemPads.get());
+}
+
+
+void CSCGEMMotherboard::retrieveGEMPads(const GEMPadDigiCollection* gemPads, unsigned id)
+{
+  auto superChamber(gem_g->superChamber(id));
+  for (auto ch : superChamber->chambers()) {
+    for (auto roll : ch->etaPartitions()) {
+      GEMDetId roll_id(roll->id());
+      auto pads_in_det = gemPads->get(roll_id);
+      for (auto pad = pads_in_det.first; pad != pads_in_det.second; ++pad) {
+        auto id_pad = std::make_pair(roll_id, *pad);
+        const int bx_shifted(lct_central_bx + pad->bx());
+        for (int bx = bx_shifted - maxDeltaBXPad_;bx <= bx_shifted + maxDeltaBXPad_; ++bx) {
+          pads_[bx].push_back(id_pad);  
+        }
+      }
+    }
+  }
+}
+
+void CSCGEMMotherboard::retrieveGEMCoPads()
+{
+  for (auto copad: gemCoPadV){
+    if (copad.first().bx() != lct_central_bx) continue;
+    coPads_[copad.bx(1)].push_back(std::make_pair(copad.roll(), copad));  
+  }
+}
+
+CSCCorrelatedLCTDigi CSCGEMMotherboard::constructLCTsGEM(const CSCALCTDigi& alct,
```

Comment:
I have been thinking about that. Although there are four different ways to construct a GEM+CSC stub: ALCT+CLCT+1 GEM, ALCT+CLCT+2 GEM, ALCT+2 GEM, CLCT+2 GEM. Hence four different functions. I'll see if I can improve it